server/src/main/java/io/crate/types/StringType.java

```diff
@@ -142,6 +155,24 @@ public String explicitCast(Object value) throws IllegalArgumentException, ClassC
         }
     }
 
+    @Override
+    public String valueForInsert(Object value) {
+        if (value == null) {
+            return null;
+        }
+        var string = convert(value);
+        if (unbound() || string.length() <= lengthLimit) {
+            return string;
+        } else {
+            if (string.substring(lengthLimit).isBlank()) {
```

Comment:
This may be fairly expensive for it does (copies the string) - to check if a region is blank.

Reply:
`substring` does not copy the string, it is just a view of the underlying char array, therefore, it is constant time complexity.

Reply:
Unfortunately not:

```
    public String substring(int beginIndex) {
        return substring(beginIndex, length());
    }

    /**
     * Returns a string that is a substring of this string. The
     * substring begins at the specified {@code beginIndex} and
     * extends to the character at index {@code endIndex - 1}.
     * Thus the length of the substring is {@code endIndex-beginIndex}.
     * <p>
     * Examples:
     * <blockquote><pre>
     * "hamburger".substring(4, 8) returns "urge"
     * "smiles".substring(1, 5) returns "mile"
     * </pre></blockquote>
     *
     * @param      beginIndex   the beginning index, inclusive.
     * @param      endIndex     the ending index, exclusive.
     * @return     the specified substring.
     * @throws     IndexOutOfBoundsException  if the
     *             {@code beginIndex} is negative, or
     *             {@code endIndex} is larger than the length of
     *             this {@code String} object, or
     *             {@code beginIndex} is larger than
     *             {@code endIndex}.
     */
    public String substring(int beginIndex, int endIndex) {
        int length = length();
        checkBoundsBeginEnd(beginIndex, endIndex, length);
        int subLen = endIndex - beginIndex;
        if (beginIndex == 0 && endIndex == length) {
            return this;
        }
        return isLatin1() ? StringLatin1.newString(value, beginIndex, subLen)
                          : StringUTF16.newString(value, beginIndex, subLen);
    }

```

```
    public static String newString(byte[] val, int index, int len) {
        return new String(Arrays.copyOfRange(val, index, index + len),
                          LATIN1);
    }
```

```
    public static String newString(byte[] val, int index, int len) {
        if (String.COMPACT_STRINGS) {
            byte[] buf = compress(val, index, len);
            if (buf != null) {
                return new String(buf, LATIN1);
            }
        }
        int last = index + len;
        return new String(Arrays.copyOfRange(val, index << 1, last << 1), UTF16);
    }

```

Reply:
indeed.. sorry, was super confident that it is not the cast :/ it seems to be different in quite old java versions.

Reply:
Or maybe confused it with `List#subList` ? I think that is just a view.

Reply:
anyways, it will be more efficient just to go through chars in the range of our interest.